Patch by Robert Scheck <robert@fedoraproject.org> for Zarafa <= 7.1.8 which implements the
POP3 STLS (STARTTLS) extension according to RFC 2595. The code is heavily based the Zarafa
IMAP gateway thus it should not introduce any new issue.

Testing: Configure zarafa-gateway for SSL/TLS usage as it would be for IMAP. However enable
also POP3S and run: "openssl s_client -connect 127.0.0.1:110 -starttls pop3". Note that the
"STLS" command does not work the connection is already using SSL/TLS (same like at IMAP).

--- zarafa-7.1.8/gateway/POP3.cpp			2014-02-20 12:55:50.000000000 +0100
+++ zarafa-7.1.8/gateway/POP3.cpp.pop3-stls		2014-03-02 23:25:29.000000000 +0100
@@ -167,7 +167,17 @@
 	strCommand = vWords[0];
 	transform(strCommand.begin(), strCommand.end(), strCommand.begin(), ::toupper);
 
-	if (strCommand.compare("USER") == 0) {
+	if (strCommand.compare("STLS") == 0) {
+		if (vWords.size() != 1) {
+			hr = HrResponse(POP3_RESP_ERR, "STLS command must have 0 arguments");
+			goto exit;
+		}
+		if (HrCmdStarttls() != hrSuccess) {
+			// log ?
+			// let the gateway quit from the socket read loop
+			hr = MAPI_E_END_OF_SESSION;
+		}
+	} else if (strCommand.compare("USER") == 0) {
 		if (vWords.size() != 2) {
 			hr = HrResponse(POP3_RESP_ERR, "User command must have 1 argument");
 			goto exit;
@@ -284,6 +294,45 @@
 }
 
 /** 
+ * @brief Handle the STLS command
+ * 
+ * Tries to set the current connection to use SSL encryption.
+ *
+ * @return hrSuccess
+ */
+HRESULT POP3::HrCmdStarttls() {
+	HRESULT hr = hrSuccess;
+
+	if (!lpChannel->sslctx()) {
+		hr = HrResponse(POP3_RESP_ERR, "STLS error in ssl context");
+		goto exit;
+	}
+
+	if (lpChannel->UsingSsl()) {
+		hr = HrResponse(POP3_RESP_ERR, "STLS already using SSL/TLS");
+		goto exit;
+	}
+
+	hr = HrResponse(POP3_RESP_OK, "Begin TLS negotiation now");
+	if (hr != hrSuccess)
+		goto exit;
+
+	hr = lpChannel->HrEnableTLS();
+	if (hr != hrSuccess) {
+		HrResponse(POP3_RESP_ERR, "Error switching to secure SSL/TLS connection");
+		lpLogger->Log(EC_LOGLEVEL_ERROR, "Error switching to SSL in STLS");
+		goto exit;
+	}
+
+	if (lpChannel->UsingSsl()) {
+		lpLogger->Log(EC_LOGLEVEL_INFO, "Using SSL now");
+	}
+
+exit:
+	return hr;
+}
+
+/** 
  * @brief Handle the USER command
  *
  * Stores the username in the class, since the password is in a second call
--- zarafa-7.1.8/gateway/POP3.h				2014-02-20 12:55:50.000000000 +0100
+++ zarafa-7.1.8/gateway/POP3.h.pop3-stls		2014-03-02 23:44:19.000000000 +0100
@@ -80,6 +80,7 @@
 	HRESULT HrDone(bool bSendResponse);
 
 private:
+	HRESULT HrCmdStarttls();
 	HRESULT HrCmdUser(const std::string &strUser);
 	HRESULT HrCmdPass(const std::string &strPass);
 	HRESULT HrCmdStat();
